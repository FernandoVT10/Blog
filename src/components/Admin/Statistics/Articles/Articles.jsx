import ArticleCard from "./ArticleCard";
import Api from "../../../../ApiController";

import { useRouter } from "next/router";
import { useState, useEffect } from "react";

import "./Articles.scss";

export default ({ setShowArticlesTable }) => {
    const [monthViews, setMonthViews] = useState([]);
    const [dayViews, setDayViews] = useState([]);

    const router = useRouter();

    useEffect(() => {
        Promise.all([
            Api.get("views/getMostViewedArticles/month/5", true),
            Api.get("views/getMostViewedArticles/day/5", true)
        ]).then(([monthViews, dayViews]) => {
            setDayViews(dayViews);
            setMonthViews(monthViews);
        });
    }, []);

    const showArticlesTable = (sort) => {
        setShowArticlesTable(true);

        router.push({
            pathname: router.pathname,
            query: { sort }
        });
    };

    return (
        <div className="statistics-articles">
            <ArticleCard
            title="Most Viewed Article This Month"
            showArticlesTable={ () => showArticlesTable("month") }
            data={monthViews}/>
            <ArticleCard
            title="Most Viewed Article This Day"
            showArticlesTable={ () => showArticlesTable("day") }
            data={dayViews}/>
        </div>
    );
};