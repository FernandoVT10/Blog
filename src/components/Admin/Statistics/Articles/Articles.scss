@use "../../../../styles/variables";

.statistics-articles {
    display: flex;

    &__card {
        width: calc(50% - 40px);
        background: variables.$primary;

        &:nth-child(1) { margin-right: 40px; }

        &-content {
            padding: 15px;
            padding-bottom: 0;
        }

        &-label {
            display: block;
            color: white;
        }

        &-views {
            display: block;
            color: white;
            font-weight: bold;
            font-size: 20px;

            &--loading {
                display: block;
                width: 100px;
                height: 30px;
                background: variables.$black;
            }
        }

        &-cover {
            margin-top: 10px;
            object-fit: cover;
            width: 100%;
            height: 200px;

            &--loading {
                display: block;
                margin-top: 10px;
                width: 100%;
                height: 200px;
                background: variables.$black;
            }
        }

        &-title {
            color: white;
            font-weight: 600;
            font-size: 22px;
            margin: 20px 15px;

            &--loading {
                margin: 20px 15px;
                display: block;
                width: 80%;
                height: 35px;
                background: variables.$black;
            }
        }
    }

    &__articles-list {
        &-item {
            position: relative;
            display: flex;
            align-items: center;
            padding: 5px 0;

            &:nth-child(2n+1) {
                background: variables.$black;
            }
        }

        &-cover {
            width: 80px;
            height: 50px;
            object-fit: cover;
        }

        &-title {
            margin-left: 15px;
            color: white;
            font-weight: bold;
            font-size: 16px;
        }

        &-views {
            position: absolute;
            right: 15px;
            color: white;
            font-weight: bold;
            font-size: 18px;
        }

        @media screen and (max-width: 576px) {
            &-item {
                padding: 20px 0;
            }

            &-cover {
                display: none;
            }
        }
    }

    &__card-footer {
        display: flex;
        justify-content: center;
        align-items: center;
        padding: 15px;
        background: variables.$black;

        &-text {
            cursor: pointer;
            color: variables.$secondary;
            font-weight: bold;
        }
    }

    @media screen and (max-width: 992px) {
        flex-direction: column;

        &__card {
            width: 100%;
            margin: 0;
            
            &:nth-child(1) { margin-bottom: 40px; }
        }
    }
}

.holi-4 {
    display: inline-block;
    width: 80px;
    height: 50px;
    background: variables.$primary;
}

.holi-5 {
    display: inline-block;
    width: 50%;
    height: 30px;
    margin-left: 15px;
    background: variables.$primary;
}

.holi-6 {
    display: inline-block;
    position: absolute;
    right: 15px;
    width: 20%;
    height: 35px;
    margin-left: 15px;
    background: variables.$primary;
}

.statistics-articles__articles-list-item:nth-child(2n) {
    .holi-4,
    .holi-5,
    .holi-6 {
        background: variables.$black;
    }
}