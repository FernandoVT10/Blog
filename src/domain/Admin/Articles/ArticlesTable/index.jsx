import Articles from "./ArticlesTable";

export default Articles;