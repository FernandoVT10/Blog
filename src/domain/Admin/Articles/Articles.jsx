import ArticlesTable from "./ArticlesTable"
import EditArticle from "./ArticleEditor/EditArticle";
import AddArticle from "./ArticleEditor/AddArticle";
import Categories from "./Categories";

import { useEffect, useState } from "react";
import { useRouter } from "next/router";

export default () => {
    const [articleId, setArticleId] = useState("");
    const [addArticle, setAddArticle] = useState(false);

    const router = useRouter();

    useEffect(() => {
        const { editArticleId } = router.query;

        if(editArticleId) {
            setArticleId(editArticleId);
        } else {
            setArticleId();
        }
    }, [router.query]);

    if(articleId) {
        return <EditArticle articleId={articleId}/>;
    }

    if(addArticle) {
        return <AddArticle setAddArticle={setAddArticle}/>;
    }

    return (
        <div>
            <Categories/>
            <ArticlesTable setAddArticle={setAddArticle}/>
        </div>
    );
}