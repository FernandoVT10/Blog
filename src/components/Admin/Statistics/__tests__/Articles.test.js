import { act } from "react-dom/test-utils";
import { render } from "react-dom";
import Articles from "../Articles/";

const MOCK_DAY_VIEWS = [
    {
        _id: 1,
        title: "Day view mock #1",
        cover: "day-cover-1.jpg",
        views: 3951
    },
    {
        _id: 2,
        title: "Day view mock #2",
        cover: "day-cover-2.jpg",
        views: 1915
    },
    {
        _id: 3,
        title: "Day view mock #1",
        cover: "day-cover-1.jpg",
        views: 1422
    }
];
const MOCK_MONTH_VIEWS = [
    {
        _id: 1,
        title: "Month view mock #1",
        cover: "month-cover-1.jpg",
        views: 3951
    },
    {
        _id: 2,
        title: "Month view mock #2",
        cover: "month-cover-2.jpg",
        views: 1915
    },
    {
        _id: 3,
        title: "Month view mock #1",
        cover: "month-cover-1.jpg",
        views: 1422
    }
];

let container;

describe('<Articles/> Component', () => {
    beforeEach(() => {
        fetchMock.doMock();

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    describe("Check if renders correctly", () => {
        it('It should render without data', async () => {
            fetchMock
            .once(JSON.stringify([]))
            .once(JSON.stringify([]));

            await act(async () => {
                render(<Articles/>, container);
            });
    
            expect(container).toMatchSnapshot();
        });

        it('It should render with data', async () => {
            fetchMock
            .once(JSON.stringify(MOCK_MONTH_VIEWS))
            .once(JSON.stringify(MOCK_DAY_VIEWS));

            await act(async () => {
                render(<Articles/>, container);
            });
    
            expect(container).toMatchSnapshot();
        });
    });
});