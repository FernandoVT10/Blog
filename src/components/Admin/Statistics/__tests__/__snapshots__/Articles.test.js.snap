// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Articles/> Component Check if renders correctly It should render with data 1`] = `
<div>
  <div
    class="statistics-articles"
  >
    <div
      class="statistics-articles__card"
    >
      <div
        class="statistics-articles__card-content"
      >
        <span
          class="statistics-articles__card-label"
        >
          Most Viewed Article This Month
        </span>
        <span
          class="statistics-articles__card-views"
        >
          3951
        </span>
        <img
          alt="Most Viewed Article This Month"
          class="statistics-articles__card-cover"
          src="/img/articles/month-cover-1.jpg"
        />
        <h3
          class="statistics-articles__card-title"
        >
          Month view mock #1
        </h3>
        <div
          class="statistics-articles__articles-list"
        >
          <div
            class="statistics-articles__articles-list-item"
          >
            <img
              alt="Article Cover"
              class="statistics-articles__articles-list-cover"
              src="/img/articles/month-cover-2.jpg"
            />
            <span
              class="statistics-articles__articles-list-title"
            >
              Month view mock #2
            </span>
            <span
              class="statistics-articles__articles-list-views"
            >
              1915
            </span>
          </div>
          <div
            class="statistics-articles__articles-list-item"
          >
            <img
              alt="Article Cover"
              class="statistics-articles__articles-list-cover"
              src="/img/articles/month-cover-1.jpg"
            />
            <span
              class="statistics-articles__articles-list-title"
            >
              Month view mock #1
            </span>
            <span
              class="statistics-articles__articles-list-views"
            >
              1422
            </span>
          </div>
        </div>
      </div>
      <div
        class="statistics-articles__card-footer"
      >
        <span
          class="statistics-articles__card-footer-text"
        >
          Show All Articles
        </span>
      </div>
    </div>
    <div
      class="statistics-articles__card"
    >
      <div
        class="statistics-articles__card-content"
      >
        <span
          class="statistics-articles__card-label"
        >
          Most Viewed Article This Day
        </span>
        <span
          class="statistics-articles__card-views"
        >
          3951
        </span>
        <img
          alt="Most Viewed Article This Day"
          class="statistics-articles__card-cover"
          src="/img/articles/day-cover-1.jpg"
        />
        <h3
          class="statistics-articles__card-title"
        >
          Day view mock #1
        </h3>
        <div
          class="statistics-articles__articles-list"
        >
          <div
            class="statistics-articles__articles-list-item"
          >
            <img
              alt="Article Cover"
              class="statistics-articles__articles-list-cover"
              src="/img/articles/day-cover-2.jpg"
            />
            <span
              class="statistics-articles__articles-list-title"
            >
              Day view mock #2
            </span>
            <span
              class="statistics-articles__articles-list-views"
            >
              1915
            </span>
          </div>
          <div
            class="statistics-articles__articles-list-item"
          >
            <img
              alt="Article Cover"
              class="statistics-articles__articles-list-cover"
              src="/img/articles/day-cover-1.jpg"
            />
            <span
              class="statistics-articles__articles-list-title"
            >
              Day view mock #1
            </span>
            <span
              class="statistics-articles__articles-list-views"
            >
              1422
            </span>
          </div>
        </div>
      </div>
      <div
        class="statistics-articles__card-footer"
      >
        <span
          class="statistics-articles__card-footer-text"
        >
          Show All Articles
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Articles/> Component Check if renders correctly It should render without data 1`] = `
<div>
  <div
    class="statistics-articles"
  >
    <div
      class="statistics-articles__card"
    >
      <div
        class="statistics-articles__card-content"
      >
        <span
          class="statistics-articles__card-label"
        >
          Most Viewed Article This Month
        </span>
        <span
          class="statistics-articles__card-views--loading"
        />
        <span
          class="statistics-articles__card-cover--loading"
        />
        <span
          class="statistics-articles__card-title--loading"
        />
        <div
          class="statistics-articles__articles-list"
        />
      </div>
      <div
        class="statistics-articles__card-footer"
      >
        <div
          class="spinner-border text-gray"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
    <div
      class="statistics-articles__card"
    >
      <div
        class="statistics-articles__card-content"
      >
        <span
          class="statistics-articles__card-label"
        >
          Most Viewed Article This Day
        </span>
        <span
          class="statistics-articles__card-views--loading"
        />
        <span
          class="statistics-articles__card-cover--loading"
        />
        <span
          class="statistics-articles__card-title--loading"
        />
        <div
          class="statistics-articles__articles-list"
        />
      </div>
      <div
        class="statistics-articles__card-footer"
      >
        <div
          class="spinner-border text-gray"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
